### PR TITLE
[SW-2280] Parse XMP

### PR DESCRIPTION
**Pull request by rugheid**

### CODE REVIEW

**Comment by rugheid** [comment]
Should I also try to parse the values? Make numbers from number strings, bools from bool string, etc?

**Comment by EvertEt** [comment]
Make sure to test the compiled version. Also try simple number and bool parsing as discussed

**Comment by rugheid** [comment]
- [x] Parse values (numbers, bool)
- [x] Test compiled version in external project

File: src/xmp-parser.js [hunk 1]
```diff
@@ -0,0 +1,462 @@
...
+        Object.keys(xmp).forEach((key) => {
+            xmp[key] = parseValue(key, xmp[key]);
+        });
+        return xmp;
+    } catch (error) {
+        return {};
+    }
+}
+
+function parseValue(key, value) {
```
Comment (rugheid) [inline src/xmp-parser.js:40]: I added the parsing here.

File: index.js [hunk 2]
```diff
@@ -103,6 +103,488 @@
 		return result
 	}
 
+	/* Based on https://github.com/mattiasw/ExifReader/blob/master/src/xmp-tags.js
```
Comment (rugheid) [inline index.js:106]: This is the build, so not necessary to review.

File: src/dom-parser.js [hunk 3]
```diff
@@ -0,0 +1,19 @@
+/* Based on https://github.com/mattiasw/ExifReader/blob/master/src/xmp-tags.js
```
Comment (rugheid) [inline src/dom-parser.js:1]: This file is just taken over.

File: src/options.mjs [hunk 4]
```diff
@@ -50,6 +52,9 @@ export const defaultOptions = {
 	// APP1 - IFD1 - Size and other information about embeded thumbnail.
 	thumbnail: false,
 
+	// XMP Parser function
+	xmpParser: XMPParser.read,
```
Comment (rugheid) [inline src/options.mjs:56]: I now parse the XMP string by default, but you could specify another parser function.

Comment (rugheid) [reply src/options.mjs:56]: This is to not divert too much from upstream.

File: src/parser.mjs [hunk 5]
```diff
@@ -613,9 +613,11 @@ export class ExifParser extends Reader {
 		if (this.options.postProcess) {
 			let start = this.xmp.indexOf('<x:xmpmeta')
 			let end = this.xmp.indexOf('x:xmpmeta>') + 10
-			this.xmp = this.xmp.slice(start, end)
+			if (start >= 0 && end > 0) {
```
Comment (rugheid) [inline src/parser.mjs:616]: This is a fix for XMP strings that don't use the `x:xmpmeta` tag, like the ones in the FLIR Duo Pro photos.

File: src/xmp-parser.js [hunk 6]
```diff
@@ -0,0 +1,462 @@
+/* Based on https://github.com/mattiasw/ExifReader/blob/master/src/xmp-tags.js
```
Comment (rugheid) [inline src/xmp-parser.js:1]: This file is taken over, except for the value parsing.

File: test/test.js [hunk 7]
```diff
@@ -292,7 +292,7 @@ describe('parser (exif data)', () => {
 	describe('Other segments', () => {
 
 		it(`should contain XMP segment (if whole file buffer is provided and options.xmp is enabled)`, async () => {
-			var exif = await parse(buffers['cookiezen.jpg'], {mergeOutput: false, xmp: true})
+			var exif = await parse(buffers['cookiezen.jpg'], {mergeOutput: false, xmp: true, xmpParser: null})
```
Comment (rugheid) [inline test/test.js:295]: Small fix to make their tests still work. We're just disabling the XMP parser.

File: test/test.js [hunk 8]
```diff
@@ -579,62 +579,85 @@ describe('parser (exif data)', () => {
...
-			assert.equal(exif.latitude, 50.85904941666667)
-			assert.equal(exif.longitude, 4.678717111111111)
+			assert.strictEqual(exif.GPSAltitudeRef, 1)
+			assert.strictEqual(exif.GPSAltitude, 116.63)
+			assert.strictEqual(exif.latitude, 50.85904941666667)
+			assert.strictEqual(exif.longitude, 4.678717111111111)
 
 			// XMP
 			assert.exists(exif.xmp)
+			assert.strictEqual(exif.xmp['rdf:about'], 'DJI Meta Data')
```
Comment (rugheid) [inline test/test.js:638]: XMP tests with parsed values.

File: src/parser.mjs [hunk 9]
```diff
@@ -613,7 +613,9 @@ export class ExifParser extends Reader {
 		if (this.options.postProcess) {
 			let start = this.xmp.indexOf('<x:xmpmeta')
 			let end = this.xmp.indexOf('x:xmpmeta>') + 10
-			this.xmp = this.xmp.slice(start, end)
+            if (start >= 0 && end > 0) {
```
Comment (EvertEt) [inline src/parser.mjs]: Minor indent

File: test/test.js [hunk 10]
```diff
@@ -579,62 +579,85 @@ describe('parser (exif data)', () => {
...
+			assert.strictEqual(exif.xmp['xmp:ModifyDate'], '2017-03-14')
+			assert.strictEqual(exif.xmp['xmp:CreateDate'], '2017-03-14')
+			assert.strictEqual(exif.xmp['tiff:Make'], 'DJI')
+			assert.strictEqual(exif.xmp['tiff:Model'], 'FC350')
+			assert.strictEqual(exif.xmp['dc:format'], 'image/jpeg')
+			assert.strictEqual(exif.xmp['drone-dji:AbsoluteAltitude'], -116.63)
+			assert.strictEqual(exif.xmp['drone-dji:RelativeAltitude'], 29.40)
```
Comment (rugheid) [inline test/test.js:645]: This is the key we actually need for this ticket.

File: src/xmp-parser.js [hunk 11]
```diff
@@ -8,17 +8,57 @@ export default {
...
+        value.value = parseValue(key, value.value);
+        return value;
+    } else if (tagTypes[key]) {
+        if (tagTypes[key] === 'bool') {
+            return ['True', 'true', '1'].includes(value);
+        } else if (tagTypes[key] === 'string') {
+            return value;
+        }
+    } else if (!isNaN(value) && value !== '') {
+        return +value;
```
Comment (EvertEt) [inline src/xmp-parser.js]: Minor: I like Number() better than implicit casting like this

File: test/test.js [hunk 12]
```diff
@@ -743,90 +800,148 @@ describe('parser (exif data)', () => {
...
+				assert.strictEqual(exif.GPSTimeStamp, '12:1:23')
+				assert.strictEqual(exif.GPSSpeedRef, 'K')
+				assert.strictEqual(exif.GPSSpeed, 0)
+				assert.strictEqual(exif.GPSTrackRef, 'T')
+				assert.strictEqual(exif.GPSTrack, 0)
+				assert.strictEqual(exif.latitude, 0)
+				assert.strictEqual(exif.longitude, 0)
 
 				// XMP
-				// TODO: Fix XMP
```
Comment (rugheid) [inline test/test.js:781]: XMP is fixed for FLIR Duo Pro

File: test/test.js [hunk 13]
```diff
@@ -743,90 +800,148 @@ describe('parser (exif data)', () => {
...
-				assert.equal(exif.latitude, 0)
-				assert.equal(exif.longitude, 0)
+				assert.strictEqual(exif.GPSAltitudeRef, 0)
+				assert.strictEqual(exif.GPSAltitude, 0)
+				assert.strictEqual(exif.GPSMapDatum, 'WGS-84')
+				assert.strictEqual(exif.latitude, 0)
+				assert.strictEqual(exif.longitude, 0)
 
 				// XMP
-				// TODO: Fix XMP
```
Comment (rugheid) [inline test/test.js:828]: XMP is fixed for FLIR Duo Pro

File: src/parser.mjs (hunk 9 above)
Comment (rugheid) [reply src/parser.mjs]: I already fixed this. :)

File: src/xmp-parser.js (hunk 11 above)
Comment (rugheid) [reply src/xmp-parser.js]: Fixed

Comment (rugheid) [reply src/xmp-parser.js]: Also did the spreading above as suggested

File: test/test.js [hunk 14]
```diff
@@ -835,290 +950,403 @@ describe('parser (exif data)', () => {
...
+			assert.strictEqual(exif.GPSStatus, 'A')
+			assert.strictEqual(exif.GPSMapDatum, 'WGS-84')
+			assert.strictEqual(exif.GPSDateStamp, '2019:03:18')
+			assert.strictEqual(exif.timestamp, '2019-03-18T11:01:20')
+			assert.strictEqual(exif.latitude, 46.59786403555556)
+			assert.strictEqual(exif.longitude, 6.6087956775)
 
 			// XMP
-			assert.exists(exif.xmp)
+            assert.exists(exif.xmp)
```
Comment (EvertEt) [inline test/test.js]: Minor: indent

**Review by EvertEt — APPROVED** [review]

File: src/xmp-parser.js [hunk 15]
```diff
@@ -0,0 +1,464 @@
...
+    if (value instanceof Array) {
+        return value.map((val) => parseValue(key, val));
+    } else if (value instanceof Object && 'value' in value) {
+        return {
+            ...value,
+            value: parseValue(key, value.value),
+        };
+    } else if (tagTypes[key]) {
+        if (tagTypes[key] === 'bool') {
+            return ['True', 'true', '1'].includes(value);
```
Comment (aaron012) [inline src/xmp-parser.js]: Test false more explicitly

**Review by aaron012 — APPROVED** [review]

File: src/xmp-parser.js (hunk 15 above)
Comment (rugheid) [reply src/xmp-parser.js]: Done

**Pull request merged**